pkg/storage/concurrency/concurrency_control.go

```diff
@@ -0,0 +1,355 @@
+//  |                     |    |--------------------------|                 |     ^
+//  |                     |    | [txn1] [txn2] [txn3] ... |                 |     |
+//  |                     |    +--------------------------+                 |     |
+//  |                     |                                                 |     |
+//  |                     +--> hold latches ---> remain at head of queues -----> evaluate ...
+//  |                                                                       |
+//  +----------+                                                            |
+//    Finish() | ---> exit wait queues ---> drop latches ----------------------> respond  ...
+//  +----------+                                                            |
+//  |                                                                       |
+//  +-----------------------------------------------------------------------+
+//
+// See the comments on individual components for a more detailed look at their
+// interface and inner-workings.
+//
+// At a high-level, requests enter the concurrency manager and immediately
+// acquire latches from the latchManager to serialize access to the keys that
+// they intend to touch. This latching takes into account the keys being
+// accessed, the MVCC timestamp of accesses, and the access method being used
+// (read vs. write) to allow for concurrency where possible. This has the effect
+// of queuing on conflicting in-flight operations until their completion.
+//
+// Once latched, the request consults the lockTable to check for any conflicting
+// locks owned by other transactions. If any are found, the request enters the
+// corresponding lockWaitQueue and its latches are dropped. The head of the
+// lockWaitQueue pushes the owner of the lock through a remote RPC that ends up
+// in the pushee's txnWaitQueue. This queue exists on the leaseholder replica of
+// the range that contains the pushee's transaction record. Other entries in the
+// queue wait for the head of the queue, eventually pushing it to detect
+// deadlocks. Once the lock is cleared, the head of the queue reacquires latches
+// and attempts to proceed while remains at the head of that lockWaitQueue to
+// ensure fairness.
+//
+// Once a request is latched and observes no conflicting locks in the lockTable
+// and no conflicting lockWaitQueues that it is not already the head of, the
+// request can proceed to evaluate. During evaluation, the request may insert or
+// remove locks from the lockTable for its own transaction. This is performed
+// transparently by a lockAwareBatch/lockAwareReadWriter. The request may also
+// need to consult its own locks to properly interpret the corresponding intents
+// in the MVCC keyspace. This is performed transparently by a lockAwareIter.
+//
+// When the request completes, it exits any lockWaitQueues that it was at the
+// head of and releases its latches. However, any locks that it inserted into
```

Comment:
Only on success, right? If the request tries to write an intent but this somehow fails below Raft, the lock should go away? Or I guess it's fine to have an in-mem lock too many, then we'll just push more than we need to (hope that doesn't confuse deadlock detection somehow). Hmm, SFU will basically mean having in-mem locks that aren't backed by replicated locks, so it will all just have to work anyway.

Reply:
> Only on success, right? If the request tries to write an intent but this somehow fails below Raft, the lock should go away?

Yes, documented.

> Hmm, SFU will basically mean having in-mem locks that aren't backed by replicated locks, so it will all just have to work anyway.

Yeah, we'll have to make it work. For now, though, I'm not going to worry about inserting locks on failure paths.